### PR TITLE
Proper build() methods for TF

### PR DESCRIPTION
**Pull request by Rocketknight1**
TensorFlow builds weights **lazily**. This means that layers do not have an `input_dim` argument and do not create weight tensors in the model `__init__()`. Instead, the layers wait until their `build()` method is called, which usually happens implicitly the first time the layer receives an input. Layers use the shape of the first input they see, or the value explicitly passed to their `build()` method, to infer their input dim and build their weight tensors.

Up until now, almost none of our TF models had explicit `build()` methods. This meant that weights were built implicitly when the model was called, which required lots of tiny hacks all over the codebase:

- We had to do an entire forward pass *inside `from_pretrained()`* to prepare the model weights so that we could load a checkpoint
- We had to be careful about call stacks and name scopes to ensure that models did not accidentally build themselves inside an existing call/name context and destroy their weight names. This meant our code had to sniff the existing TF call stack, which (among many other issues) completely breaks Keras 3.
- Several models had explicit calls to `tf.name_scope()` inside their forward pass (!) to control their weight names, which only worked because the weights were always built there

This had always been a big chunk of tech debt that I'd wanted to fix, but it was such a large task that I never really had time. However, with Keras 3 approaching, it became quite urgent. I tried getting GPT-4 to figure out the `build()` shapes automatically, but it generally failed, so I had to resort to using `ast` and static analysis of the PyTorch and TF modeling files to cross-match layers from TF to PyTorch code, using the input size arguments from PyTorch to automatically create and populate new `build()` methods, and then did a manual pass afterwards to fix up the remaining issues.

As a result, after this PR:

- All models now have correct `build()` methods
- Weight names are correct even if models are called weirdly, because we can now tightly control the `build()` hierarchy
- Probably the single biggest source of TF bugs we had is gone
- No more forward passes when building models, including with `from_pretrained()`! Should make model loading significantly faster, especially on CPU, and should help in the CI.
- A major Keras 3 blocker is removed.

While I was working on this PR, I also encountered some other issues that I fixed in passing:
- Added a `build_in_name_scope()` method and refactored some tests/methods to use it instead. Calling this method yields the same name hierarchy as implicitly calling `build()` when doing a forward pass, whereas directly calling `model.build()` does not (because TF enters a name_scope in `__call__()`)
- Updated TFAdaptivePool2D for Data2Vec, should massively improve model performance
- Fixed some details in the `TFSequenceSummary` and `TFConv1D` classes. These are mostly used by older models.

**Note to reviewers:** Most of this PR was generated automatically, and just consists of walls of new `build()` methods. You can generally trust that these methods are correct so long as the CI is green, so you hopefully don't have to read them all - there's >11,000 lines of them! The main things to review are the changes in core files like `modeling_tf_utils.py`, the new `build_in_name_scope()` method, etc.

### CODE REVIEW

File: tests/pipelines/test_pipelines_summarization.py [hunk 1]
```diff
@@ -67,8 +67,8 @@ def run_pipeline_test(self, summarizer, _):
             # the embedding layer.
             if not (
                 isinstance(model, TFPreTrainedModel)
-                and get_gpu_count() > 0
                 and len(summarizer.model.trainable_weights) > 0
+                and "GPU" in summarizer.model.trainable_weights[0].device
```
Comment (Rocketknight1) [inline tests/pipelines/test_pipelines_summarization.py:71]: Quick note to any reviewers so they can make sense of this one: The problem is that `get_gpu_count()` is supposed to be framework-neutral, but actually checks the frameworks in order using `is_torch_available()` etc. and returns the GPU count for the first framework that matches.

This is very risky for TensorFlow, because TF environments will often have Torch as well, and if Torch is present then the *Torch* GPU count is returned instead, which may be 0 even if TF is running on GPU. 

I just refactored the check to not use that function.

**Comment by Rocketknight1** [comment]
oh my god the tests pass i didn't think this was ever going to happen

**Review by gante — APPROVED** [review]
Nice and clean 🔥 I can't imagine the time you took to find this "simple" solution 👀

**Comment by Rocketknight1** [comment]
Thanks! There's one thing left to add - some layers are buildable with int shapes in Keras 2, but that always fails in Keras 3. I'm going to do a quick replacement so that those become actual shapes (with extra `None` dimensions) - it should have no impact on Keras 2 either way.

**Comment by Rocketknight1** [comment]
Quick update: The build shapes all have proper ranks instead of just being ints now, but our old method of controlling names with `tf.name_scope()` isn't working for Keras 3 - I've asked Chollet what the recommended solution there is

**Comment by Rocketknight1** [comment]
Got a solution, but I think it fits better in another PR! I'm gonna merge this one for now and see what shakes out in the nightly CI, while I work on the next phase of Keras 3 compatibility.

**Comment by ArthurZucker** [comment]
Not sure I get why this was merged?

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -3450,6 +3449,14 @@ def call(self, inputs, cls_index=None, training=False):
 
         return output
 
+    def build(self, input_shape):
+        if self.built:
+            return
+        self.built = True
+        if getattr(self, "summary", None) is not None:
```
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:3456]: What's the significance of "summary" here?

File: src/transformers/models/albert/modeling_tf_albert.py [hunk 3]
```diff
@@ -146,7 +146,7 @@ def __init__(self, config: AlbertConfig, **kwargs):
         self.LayerNorm = tf.keras.layers.LayerNormalization(epsilon=config.layer_norm_eps, name="LayerNorm")
         self.dropout = tf.keras.layers.Dropout(rate=config.hidden_dropout_prob)
 
-    def build(self, input_shape: tf.TensorShape):
+    def build(self, input_shape=None):
```
Comment (amyeroberts) [inline src/transformers/models/albert/modeling_tf_albert.py:149]: Nit - shouldn't this become an optional? 

```suggestion
    def build(self, input_shape: Optional[tf.TensorShape] = None):
```

File: src/transformers/models/albert/modeling_tf_albert.py [hunk 4]
```diff
@@ -246,6 +251,7 @@ def __init__(self, config: AlbertConfig, **kwargs):
         # Two different dropout probabilities; see https://github.com/google-research/albert/blob/master/modeling.py#L971-L993
         self.attention_dropout = tf.keras.layers.Dropout(rate=config.attention_probs_dropout_prob)
         self.output_dropout = tf.keras.layers.Dropout(rate=config.hidden_dropout_prob)
+        self.config = config
```
Comment (amyeroberts) [inline src/transformers/models/albert/modeling_tf_albert.py:254]: Storing the whole config in each layer can start to make things a lot bigger if it's large e.g. `id2label` with many classes. In the case of TF with safetensors, are we just instantiating the class from the library and loading the weight e.g. we don't store all the class attributes when saving?

File: src/transformers/models/camembert/modeling_tf_camembert.py [hunk 5]
```diff
@@ -965,10 +1086,18 @@ def __init__(self, config, input_embeddings, **kwargs):
         # an output-only bias for each token.
         self.decoder = input_embeddings
 
-    def build(self, input_shape):
+    def build(self, input_shape=None):
         self.bias = self.add_weight(shape=(self.config.vocab_size,), initializer="zeros", trainable=True, name="bias")
```
Comment (amyeroberts) [inline src/transformers/models/camembert/modeling_tf_camembert.py:1090]: Shouldn't this go after the check of `if self.built`?

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 6]
```diff
@@ -169,7 +169,12 @@ def build(self, input_shape: tf.TensorShape = None):
                 name="embeddings",
             )
 
-        super().build(input_shape)
+        if self.built:
```
Comment (amyeroberts) [inline src/transformers/models/clip/modeling_tf_clip.py:172]: Same here - can we return quickly without calling `self.add_weight` if `self.built` is True?

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 7]
```diff
@@ -766,7 +877,21 @@ def build(self, input_shape: tf.TensorShape = None):
             name="logit_scale",
         )
 
-        super().build(input_shape)
+        if self.built:
```
Comment (amyeroberts) [inline src/transformers/models/clip/modeling_tf_clip.py:880]: Same q here. I'm not going to message any time I see it in case it's nothing but if it's not then you'll have to go through the PR to find all the cases!

File: src/transformers/models/hubert/modeling_tf_hubert.py [hunk 8]
```diff
@@ -416,11 +416,6 @@ def _normalize_kernel(self):
 
     def build(self, input_shape):
         if not self.built:
-            input_shape = input_shape.as_list()
-            # If a specific input shape is passed in, we need to modify it to account for padding
-            # Not necessary if those portions of the shape are None
-            if input_shape[-2] is not None:
-                input_shape[-2] += self.explicit_padding * 2
             super().build(input_shape)
```
Comment (amyeroberts) [inline src/transformers/models/hubert/modeling_tf_hubert.py:419]: Do we still want this super().build() call here?

**Review by amyeroberts — COMMENTED** [review]
As promised, a quick follow up review :) Thanks for this massive piece of work! Layer naming / building is definitely easier to follow like this ❤️ 

Just a few comments, mostly nits. Main ones are about saving the config in all of the layer modules and where the early return happens in some methods.

File: src/transformers/models/albert/modeling_tf_albert.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/models/albert/modeling_tf_albert.py:254]: This is a good point, but I think it's okay! Our `save_pretrained` methods should just save the weights, and not save attributes of the layers like this. Also, setting `self.config = config` just creates a reference to the same underlying `config` object, so it doesn't use extra memory when the model is initialized either.

Comment (Rocketknight1) [reply src/transformers/models/albert/modeling_tf_albert.py:254]: In general, a lot of our models do `self.config = config` in the `__init__` anyway! I just needed to add a lot more of it so that the `build()` methods could see the config vars they need.

File: src/transformers/models/camembert/modeling_tf_camembert.py (hunk 5 above)
Comment (Rocketknight1) [reply src/transformers/models/camembert/modeling_tf_camembert.py:1090]: Probably! In general, this happens when the class had an existing build method, in which case my new method is appended to the end of it. Existing build methods in the codebase don't always have an `if built:` clause. It shouldn't actually cause too many problems, but I could consider a pass to fix it up if it does.

File: src/transformers/models/hubert/modeling_tf_hubert.py (hunk 8 above)
Comment (Rocketknight1) [reply src/transformers/models/hubert/modeling_tf_hubert.py:419]: Probably not, but it should be harmless - I don't even think the `super()` method really does anything in TF anymore!
